src/java_bytecode/java_object_factory.cpp

```diff
@@ -494,6 +496,167 @@ class recursion_set_entryt
   }
 };
 
+/// Get max value for an integral type
+/// \param type:
+///   Type to find maximum value for
+/// \return Maximum integral value
+static size_t max_value(const typet& type)
+{
+  if(type.id()==ID_signedbv)
+    return std::numeric_limits<int32_t>::max();
+  else if(type.id()==ID_unsignedbv)
+    return std::numeric_limits<uint32_t>::max();
+  UNREACHABLE;
+}
+
+/// Initialize a nondeterministic String structure
+/// \param obj: struct to initialize, must have been declared using
+///   code of the form:
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// struct java.lang.String { struct \@java.lang.Object;
+///   int length; char *data; } tmp_object_factory$1;
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// \param max_nondet_string_length: maximum length of strings to initialize
+/// \param loc: location in the source
+/// \param symbol_table: the symbol table
+/// \return code for initialization of the strings
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// int tmp_object_factory$1;
+/// tmp_object_factory$1 = NONDET(int);
+/// __CPROVER_assume(tmp_object_factory$1 >= 0);
+/// __CPROVER_assume(tmp_object_factory$1 <= 20);
+/// char nondet_infinite_array$2[INFINITY()];
+/// nondet_infinite_array$2 = NONDET(char [INFINITY()]);
+/// int return_array$3;
+/// return_array$3 = cprover_associate_array_to_pointer_func
+///   (nondet_infinite_array$2, &nondet_infinite_array$2[0]);
+/// int return_array$4;
+/// return_array$4 = cprover_associate_length_to_array_func
+///   (nondet_infinite_array$2, tmp_object_factory$1);
+/// arg = { .\@java.lang.Object={ .\@class_identifier="java.lang.String",
+///   .\@lock=false }, .length=tmp_object_factory$1,
+///   .data=nondet_infinite_array$2 };
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// Unit tests in `unit/java_bytecode/java_object_factory/` ensure
+/// it is the case.
+codet initialize_nondet_string_struct(
+  const exprt &obj,
+  const std::size_t &max_nondet_string_length,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table)
+{
+  PRECONDITION(
+    java_string_library_preprocesst::implements_java_char_sequence(obj.type()));
+
+  const namespacet ns(symbol_table);
+  code_blockt code;
+
+  // `obj` is `*expr`
+  const struct_typet struct_type=
+    obj.type().id()==ID_symbol?
+    to_struct_type(ns.follow(obj.type())):
+    to_struct_type(obj.type());
+  const irep_idt &class_id=struct_type.get_tag();
+
+
+  // @clsid = String and @lock = false:
+  const symbol_typet jlo_symbol("java::java.lang.Object");
+  const struct_typet jlo_type=to_struct_type(ns.follow(jlo_symbol));
+  struct_exprt jlo_init(jlo_symbol);
+  const constant_exprt class_id_expr(class_id, jlo_type.components()[0].type());
```

Comment:
I find it strange that we have to access type components in that way. Aren't there any accessors?